pkg/agent/controller/networkpolicy/fqdn_test.go

```diff
@@ -584,3 +587,151 @@ func TestSyncDirtyRules(t *testing.T) {
 		})
 	}
 }
+
+func TestOnDNSResponse(t *testing.T) {
+	testFQDN := "fqdn-test-pod.lfx.test"
+	selectorItem1 := fqdnSelectorItem{
+		matchName: testFQDN,
+	}
+	selectorItem2 := fqdnSelectorItem{
+		matchName: "random-domain.com",
+	}
+	currentTime := time.Now()
+
+	tests := []struct {
+		name                  string
+		existingDNSCache      map[string]dnsMeta
+		dnsResponseIPs        map[string]ipWithExpiration
+		expectedIPs           map[string]ipWithExpiration
+		expectedRequeryAfter  *time.Duration
+		mockSelectorToRuleIDs map[fqdnSelectorItem]sets.Set[string]
+	}{
+		{
+			name: "new IP added",
+			existingDNSCache: map[string]dnsMeta{
+				testFQDN: {
+					responseIPs: map[string]ipWithExpiration{
+						"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+						"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithExpiration{
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(10 * time.Second)},
+			},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+				"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(10 * time.Second)},
+			},
+			expectedRequeryAfter: ptr.To(5 * time.Second),
+		},
+		{
+			name: "existing DNS cache not impacted",
+			existingDNSCache: map[string]dnsMeta{
+				testFQDN: {
+					responseIPs: map[string]ipWithExpiration{
+						"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+						"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithExpiration{},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+				"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(6 * time.Second)},
+			},
+		},
+		{
+			name: "old IP present in DNS response is retained with an updated TTL fetched from response",
```

Comment:
I don't understand why we do not expect a requeue for this test case (`expectedRequeryAfter` is `nil`).
A requeue is pretty much needed every time (except if there is no cache entry and the DNS response doesn't match an FQDN selector, which is your last test case. This is different from needing to sync the rules (which unfortunately we don't check for in the test yet, we can do it as a follow-up PR potentially). cc @Dyanngg

Reply:
I think this is right. Even if the IPs remain the same, as long as we have rules in place and an entry in the DNS cache, we should keep sending DNS requests by queuing them.